{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import xlrd\n",
    "import numpy as np\n",
    "import pandas\n",
    "from pandas import DataFrame\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# import data from excel \n",
    "if __name__ == '__main__' :\n",
    "    book = xlrd.open_workbook ('C:/research/python/data/RNA_Bric_seq_final.xlsx')\n",
    "    sheet = book.sheet_by_name('HeLa')\n",
    "    data = []\n",
    "    data = [[sheet.cell_value(r, col)\n",
    "                for col in range(sheet.ncols)]\n",
    "                for r in range(sheet.nrows)]\n",
    "\n",
    "# numbers are registered as numbers, while leaving the strings alone\n",
    "for i in range (len (data)):\n",
    "    for j in range(len(data[0])):\n",
    "        try:\n",
    "            data[i][j] = float (data[i][j])\n",
    "        except:\n",
    "            pass\n",
    "        \n",
    "# Wrap the data in a pandas DataFrame\n",
    "df = DataFrame(data[1:], columns = data[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Wrap the data in a pandas DataFrame\n",
    "df = DataFrame(data[1:], columns = data[0])\n",
    "\n",
    "# extract the significant increased genes by siPUM1\n",
    "df_q = df[df['q_value'] < 0.05]\n",
    "df_q = df[df['q_value'] < 0.05]\n",
    "df_qup = df_q[df_q['log2_change_rpkm'] > 0]\n",
    "df_p = df_qup[df_qup['P_Welch'] < 0.05]\n",
    "df_pup = df_p[df_p['log2_change_half_life'] > 0]\n",
    "df_up = df_pup[df_pup['rpkm_siCTRL'] > 1.0]\n",
    "\n",
    "# Define the data set of which genes contain PUM motifs\n",
    "df_up_motif = df_up[df_up['number'] > 0]\n",
    "\n",
    "# Define the basic transcripts set : RPKM is larger than 0.1 & t_half > 0\n",
    "df_rpkm = df[df['rpkm_siCTRL'] > 0.1]\n",
    "df_base = df_rpkm[df_rpkm['half_life_siCTRL'] > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# import data from excel \n",
    "if __name__ == '__main__' :\n",
    "    book = xlrd.open_workbook ('C:/research/python/data/random_conservation.xlsx')\n",
    "    sheet = book.sheet_by_name('Sheet1')\n",
    "    data2 = []\n",
    "    data2 = [[sheet.cell_value(r, col)\n",
    "                for col in range(sheet.ncols)]\n",
    "                for r in range(sheet.nrows)]\n",
    "\n",
    "# numbers are registered as numbers, while leaving the strings alone\n",
    "for i in range (len (data2)):\n",
    "    for j in range(len(data2[0])):\n",
    "        try:\n",
    "            data2[i][j] = float (data2[i][j])\n",
    "        except:\n",
    "            pass\n",
    "        \n",
    "# Wrap the data in a pandas DataFrame\n",
    "df_ran = DataFrame(data2[1:], columns = data2[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df_ran1 = df_ran.dropna(subset=['Conservation_score'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df1 = df.dropna(subset=['Conservation_score'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# convert dataframe to numpy array\n",
    "df_con = df1['Conservation_score']\n",
    "df_ran_con = df_ran1['Conservation_score']\n",
    "\n",
    "con_PUM = df_con.as_matrix()\n",
    "con_Ran = df_ran_con.as_matrix()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "Conserve_PUM = np.array([])\n",
    "Conserve_ran = np.array([])\n",
    "\n",
    "N = np.count_nonzero(con_PUM)\n",
    "M = np.count_nonzero(con_Ran)\n",
    "\n",
    "for i in range(N):\n",
    "    if type(con_PUM[i]) == float:                    # 3'UTR contain only 1 motif\n",
    "        Conserve_PUM = np.append(Conserve_PUM, con_PUM[i])\n",
    "    elif type(con_PUM[i]) != float:                 # 3'UTR contains multiple motifs\n",
    "        conP1 = con_PUM[i].split(\"|\")\n",
    "        conP2 = [float(j) for j in conP1]\n",
    "        for j in range(len(conP2)):\n",
    "            Conserve_PUM = np.append(Conserve_PUM, conP2)\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "for i in range(M):\n",
    "    conR1 = con_Ran[i].split(\"|\")\n",
    "    conR2 = [float(j) for j in conR1]\n",
    "    for j in range(len(conR2)):\n",
    "        Conserve_ran = np.append(Conserve_ran, conR2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "plt.hist(Conserve_PUM, bins = 200, alpha = 0.8, color = 'c', normed = True, label = 'PUM_motif')\n",
    "plt.hist(Conserve_ran, bins = 200, alpha = 0.5, color = 'k', normed = True, label = 'Random sequence')\n",
    "\n",
    "plt.legend()\n",
    "plt.xlabel(\"Conservation_score (phyloP)\")\n",
    "plt.ylabel(\"Frequency\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_up1 = df_up.dropna(subset=['Conservation_score'])\n",
    "df_up_con = df_up1['Conservation_score']\n",
    "\n",
    "con_PUM1 = df_up_con.as_matrix()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Conserve_PUM1 = np.array([])\n",
    "\n",
    "N1 = np.count_nonzero(con_PUM1)\n",
    "\n",
    "for i in range(N1):\n",
    "    if type(con_PUM1[i]) == float:                    # 3'UTR contain only 1 motif\n",
    "        Conserve_PUM1 = np.append(Conserve_PUM1, con_PUM1[i])\n",
    "    elif type(con_PUM1[i]) != float:                 # 3'UTR contains multiple motifs\n",
    "        conPum1 = con_PUM1[i].split(\"|\")\n",
    "        conPum2 = [float(j) for j in conPum1]\n",
    "        for j in range(len(conPum2)):\n",
    "            Conserve_PUM1 = np.append(Conserve_PUM1, conPum2)\n",
    "    else:\n",
    "        pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "plt.hist(Conserve_PUM, bins = 200, alpha = 0.5, color = 'g', normed = True, label = 'PUM_motif')\n",
    "plt.hist(Conserve_PUM1, bins = 200, alpha = 0.5, color = 'r', normed = True, label = 'PUM1 target')\n",
    "\n",
    "plt.legend()\n",
    "plt.xlabel(\"Conservation score\")\n",
    "plt.ylabel(\"Frequency\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fig = plt.figure()\n",
    "ax = fig.add_subplot(111)\n",
    "\n",
    "ax.boxplot([Conserve_ran, Conserve_PUM, Conserve_PUM1], notch = True)\n",
    "\n",
    "# customize the whiskers, caps, fliers and medians\n",
    "plt.ylabel('Conservation_score (phyloP)')   \n",
    "plt.xticks([1,2,3], ['Random sequence','PUM motif in all gene', 'PUM motif in PUM1 target'])\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "plt.figure();\n",
    "plt.subplot(2,1,1)\n",
    "plt.hist(Conserve_PUM, bins = 200, alpha = 0.5, color = 'c', normed = True, label = 'PUM_motif')\n",
    "plt.legend(prop={'size':15})\n",
    "plt.ylabel('Frequency')\n",
    "plt.xlim(-2,6)\n",
    "\n",
    "plt.subplot(2,1,2)\n",
    "plt.hist(Conserve_ran, bins = 200, alpha = 0.5, color = 'k', normed = True, label = 'Random sequence')\n",
    "plt.legend(prop={'size':15})\n",
    "plt.xlabel('Conservation_score (phyloP)')\n",
    "plt.ylabel('Frequency')\n",
    "plt.xlim(-2,6)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df2 = df1.loc[:,['symbol', 'number', 'sequence', 'position', '3UTR_length', 'Conservation_score']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df3 = df2.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>number</th>\n",
       "      <th>sequence</th>\n",
       "      <th>position</th>\n",
       "      <th>3UTR_length</th>\n",
       "      <th>Conservation_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>4</td>\n",
       "      <td>TGTACATT|TGTAAATA|TGTATATA|TGTAAATT</td>\n",
       "      <td>1760|4226|4334|6695</td>\n",
       "      <td>7320</td>\n",
       "      <td>0.04600000000000001|0.773|0.12349999999999998|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AADACL3</td>\n",
       "      <td>1</td>\n",
       "      <td>TGTACATA</td>\n",
       "      <td>2444</td>\n",
       "      <td>2763</td>\n",
       "      <td>0.00375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AADAT</td>\n",
       "      <td>1</td>\n",
       "      <td>TGTAAAAA</td>\n",
       "      <td>407</td>\n",
       "      <td>705</td>\n",
       "      <td>0.190875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAED1</td>\n",
       "      <td>1</td>\n",
       "      <td>TGTATATA</td>\n",
       "      <td>438</td>\n",
       "      <td>507</td>\n",
       "      <td>0.459625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAK1</td>\n",
       "      <td>8</td>\n",
       "      <td>TGTAAATT|TGTATAAA|TGTACAAA|TGTATAAA|TGTATAAA|T...</td>\n",
       "      <td>972|981|1593|2054|4873|4967|6895|14827</td>\n",
       "      <td>17874</td>\n",
       "      <td>2.6399999999999997|2.553125|1.105|1.9307500000...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    symbol  number                                           sequence  \\\n",
       "0     A1CF       4                TGTACATT|TGTAAATA|TGTATATA|TGTAAATT   \n",
       "1  AADACL3       1                                           TGTACATA   \n",
       "2    AADAT       1                                           TGTAAAAA   \n",
       "3    AAED1       1                                           TGTATATA   \n",
       "4     AAK1       8  TGTAAATT|TGTATAAA|TGTACAAA|TGTATAAA|TGTATAAA|T...   \n",
       "\n",
       "                                 position  3UTR_length  \\\n",
       "0                     1760|4226|4334|6695         7320   \n",
       "1                                    2444         2763   \n",
       "2                                     407          705   \n",
       "3                                     438          507   \n",
       "4  972|981|1593|2054|4873|4967|6895|14827        17874   \n",
       "\n",
       "                                  Conservation_score  \n",
       "0  0.04600000000000001|0.773|0.12349999999999998|...  \n",
       "1                                            0.00375  \n",
       "2                                           0.190875  \n",
       "3                                           0.459625  \n",
       "4  2.6399999999999997|2.553125|1.105|1.9307500000...  "
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df3.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "Count = len(df3)\n",
    "\n",
    "dic = {}\n",
    "\n",
    "gene = df3.iloc[0,0]\n",
    "num = df3.iloc[0,1]\n",
    "seq = df3.iloc[0,2].split('|')\n",
    "pos = df3.iloc[0,3].split('|')\n",
    "length = df3.iloc[0,4]\n",
    "con = df3.iloc[0,3].split('|')\n",
    "\n",
    "for i in range (int(num)):\n",
    "    gene_name = gene +'_'+ str(i)\n",
    "    dic[gene_name] = [seq[i], int(pos[i]), int(length), int(con[i])]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "Count = len(df3)\n",
    "\n",
    "dic = {}\n",
    "\n",
    "for j in range (int(Count)):\n",
    "    \n",
    "    gene = df3.iloc[j,0]\n",
    "    num = df3.iloc[j,1]\n",
    "    if num == 1:\n",
    "        seq = df3.iloc[j,2]\n",
    "        pos = df3.iloc[j,3]\n",
    "        length = df3.iloc[j,4]\n",
    "        con = df3.iloc[j,5]\n",
    "        gene_name = gene\n",
    "        dic[gene_name] = [seq, int(pos), int(length), float(con)]\n",
    "    elif num >1:\n",
    "        seq = df3.iloc[j,2].split('|')\n",
    "        pos = df3.iloc[j,3].split('|')\n",
    "        length = df3.iloc[j,4]\n",
    "        con = df3.iloc[j,5].split('|')\n",
    "\n",
    "        for i in range (int(num)):\n",
    "            gene_name = str(gene) +'_'+ str(i)\n",
    "            dic[gene_name] = [seq[i], int(pos[i]), int(length), float(con[i])]\n",
    "    else:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PUM motif</th>\n",
       "      <th>position</th>\n",
       "      <th>UTR length</th>\n",
       "      <th>phyloP</th>\n",
       "      <th>distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42434.0</th>\n",
       "      <td>TGTAAATT</td>\n",
       "      <td>469</td>\n",
       "      <td>2757</td>\n",
       "      <td>0.816625</td>\n",
       "      <td>2288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42615.0</th>\n",
       "      <td>TGTAAAAA</td>\n",
       "      <td>1785</td>\n",
       "      <td>2089</td>\n",
       "      <td>-0.185750</td>\n",
       "      <td>304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42620.0</th>\n",
       "      <td>TGTAAAAA</td>\n",
       "      <td>977</td>\n",
       "      <td>1238</td>\n",
       "      <td>2.995500</td>\n",
       "      <td>261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42622.0</th>\n",
       "      <td>TGTATAAA</td>\n",
       "      <td>1872</td>\n",
       "      <td>1938</td>\n",
       "      <td>2.347500</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42623.0</th>\n",
       "      <td>TGTATAAA</td>\n",
       "      <td>354</td>\n",
       "      <td>1470</td>\n",
       "      <td>-0.187000</td>\n",
       "      <td>1116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42628.0</th>\n",
       "      <td>TGTAAAAA</td>\n",
       "      <td>848</td>\n",
       "      <td>901</td>\n",
       "      <td>-0.297750</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42430.0_0</th>\n",
       "      <td>TGTATATA</td>\n",
       "      <td>1928</td>\n",
       "      <td>4450</td>\n",
       "      <td>1.293000</td>\n",
       "      <td>2522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42430.0_1</th>\n",
       "      <td>TGTATAAA</td>\n",
       "      <td>2056</td>\n",
       "      <td>4450</td>\n",
       "      <td>0.122375</td>\n",
       "      <td>2394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42430.0_2</th>\n",
       "      <td>TGTATATT</td>\n",
       "      <td>2115</td>\n",
       "      <td>4450</td>\n",
       "      <td>0.320750</td>\n",
       "      <td>2335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42430.0_3</th>\n",
       "      <td>TGTATATT</td>\n",
       "      <td>4002</td>\n",
       "      <td>4450</td>\n",
       "      <td>2.536500</td>\n",
       "      <td>448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42435.0_0</th>\n",
       "      <td>TGTAAATA</td>\n",
       "      <td>209</td>\n",
       "      <td>4354</td>\n",
       "      <td>2.247125</td>\n",
       "      <td>4145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42435.0_1</th>\n",
       "      <td>TGTATATT</td>\n",
       "      <td>282</td>\n",
       "      <td>4354</td>\n",
       "      <td>3.640250</td>\n",
       "      <td>4072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42435.0_2</th>\n",
       "      <td>TGTACATT</td>\n",
       "      <td>440</td>\n",
       "      <td>4354</td>\n",
       "      <td>0.154875</td>\n",
       "      <td>3914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42435.0_3</th>\n",
       "      <td>TGTACATT</td>\n",
       "      <td>596</td>\n",
       "      <td>4354</td>\n",
       "      <td>2.631625</td>\n",
       "      <td>3758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42435.0_4</th>\n",
       "      <td>TGTATATA</td>\n",
       "      <td>4071</td>\n",
       "      <td>4354</td>\n",
       "      <td>0.107625</td>\n",
       "      <td>283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42436.0_0</th>\n",
       "      <td>TGTATATA</td>\n",
       "      <td>142</td>\n",
       "      <td>1241</td>\n",
       "      <td>0.847875</td>\n",
       "      <td>1099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42436.0_1</th>\n",
       "      <td>TGTATAAA</td>\n",
       "      <td>616</td>\n",
       "      <td>1241</td>\n",
       "      <td>0.847500</td>\n",
       "      <td>625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42436.0_2</th>\n",
       "      <td>TGTAAATT</td>\n",
       "      <td>714</td>\n",
       "      <td>1241</td>\n",
       "      <td>1.211750</td>\n",
       "      <td>527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42621.0_0</th>\n",
       "      <td>TGTATAAA</td>\n",
       "      <td>619</td>\n",
       "      <td>2815</td>\n",
       "      <td>1.121125</td>\n",
       "      <td>2196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42621.0_1</th>\n",
       "      <td>TGTATATA</td>\n",
       "      <td>976</td>\n",
       "      <td>2815</td>\n",
       "      <td>-0.286875</td>\n",
       "      <td>1839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42627.0_0</th>\n",
       "      <td>TGTATATT</td>\n",
       "      <td>113</td>\n",
       "      <td>2368</td>\n",
       "      <td>0.196375</td>\n",
       "      <td>2255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42627.0_1</th>\n",
       "      <td>TGTATATT</td>\n",
       "      <td>2106</td>\n",
       "      <td>2368</td>\n",
       "      <td>0.037750</td>\n",
       "      <td>262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF_0</th>\n",
       "      <td>TGTACATT</td>\n",
       "      <td>1760</td>\n",
       "      <td>7320</td>\n",
       "      <td>0.046000</td>\n",
       "      <td>5560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF_1</th>\n",
       "      <td>TGTAAATA</td>\n",
       "      <td>4226</td>\n",
       "      <td>7320</td>\n",
       "      <td>0.773000</td>\n",
       "      <td>3094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF_2</th>\n",
       "      <td>TGTATATA</td>\n",
       "      <td>4334</td>\n",
       "      <td>7320</td>\n",
       "      <td>0.123500</td>\n",
       "      <td>2986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF_3</th>\n",
       "      <td>TGTAAATT</td>\n",
       "      <td>6695</td>\n",
       "      <td>7320</td>\n",
       "      <td>0.447250</td>\n",
       "      <td>625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AADACL3</th>\n",
       "      <td>TGTACATA</td>\n",
       "      <td>2444</td>\n",
       "      <td>2763</td>\n",
       "      <td>0.003750</td>\n",
       "      <td>319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AADAT</th>\n",
       "      <td>TGTAAAAA</td>\n",
       "      <td>407</td>\n",
       "      <td>705</td>\n",
       "      <td>0.190875</td>\n",
       "      <td>298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAED1</th>\n",
       "      <td>TGTATATA</td>\n",
       "      <td>438</td>\n",
       "      <td>507</td>\n",
       "      <td>0.459625</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAK1_0</th>\n",
       "      <td>TGTAAATT</td>\n",
       "      <td>972</td>\n",
       "      <td>17874</td>\n",
       "      <td>2.640000</td>\n",
       "      <td>16902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZRANB2</th>\n",
       "      <td>TGTAAATA</td>\n",
       "      <td>1817</td>\n",
       "      <td>1888</td>\n",
       "      <td>2.950750</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZRANB3</th>\n",
       "      <td>TGTAAATA</td>\n",
       "      <td>1377</td>\n",
       "      <td>3373</td>\n",
       "      <td>0.303375</td>\n",
       "      <td>1996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN12</th>\n",
       "      <td>TGTAAATT</td>\n",
       "      <td>320</td>\n",
       "      <td>1503</td>\n",
       "      <td>0.400500</td>\n",
       "      <td>1183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN20</th>\n",
       "      <td>TGTATATA</td>\n",
       "      <td>35</td>\n",
       "      <td>919</td>\n",
       "      <td>-0.112500</td>\n",
       "      <td>884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN23</th>\n",
       "      <td>TGTAAATA</td>\n",
       "      <td>1572</td>\n",
       "      <td>1810</td>\n",
       "      <td>0.383750</td>\n",
       "      <td>238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN25</th>\n",
       "      <td>TGTAAATA</td>\n",
       "      <td>1620</td>\n",
       "      <td>2387</td>\n",
       "      <td>0.473125</td>\n",
       "      <td>767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN30</th>\n",
       "      <td>TGTATATA</td>\n",
       "      <td>1174</td>\n",
       "      <td>2392</td>\n",
       "      <td>-0.312250</td>\n",
       "      <td>1218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN4_0</th>\n",
       "      <td>TGTATAAA</td>\n",
       "      <td>19</td>\n",
       "      <td>247</td>\n",
       "      <td>-0.209250</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN4_1</th>\n",
       "      <td>TGTATATT</td>\n",
       "      <td>41</td>\n",
       "      <td>247</td>\n",
       "      <td>-0.239000</td>\n",
       "      <td>206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSWIM1</th>\n",
       "      <td>TGTATATT</td>\n",
       "      <td>273</td>\n",
       "      <td>1216</td>\n",
       "      <td>0.265000</td>\n",
       "      <td>943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSWIM2</th>\n",
       "      <td>TGTATATA</td>\n",
       "      <td>88</td>\n",
       "      <td>504</td>\n",
       "      <td>0.219375</td>\n",
       "      <td>416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSWIM4</th>\n",
       "      <td>TGTATATT</td>\n",
       "      <td>998</td>\n",
       "      <td>1180</td>\n",
       "      <td>1.156625</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSWIM5_0</th>\n",
       "      <td>TGTAAATA</td>\n",
       "      <td>371</td>\n",
       "      <td>2050</td>\n",
       "      <td>2.706125</td>\n",
       "      <td>1679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSWIM5_1</th>\n",
       "      <td>TGTACATA</td>\n",
       "      <td>900</td>\n",
       "      <td>2050</td>\n",
       "      <td>-0.100125</td>\n",
       "      <td>1150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSWIM5_2</th>\n",
       "      <td>TGTATAAA</td>\n",
       "      <td>1121</td>\n",
       "      <td>2050</td>\n",
       "      <td>0.233500</td>\n",
       "      <td>929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSWIM6_0</th>\n",
       "      <td>TGTATATA</td>\n",
       "      <td>645</td>\n",
       "      <td>1855</td>\n",
       "      <td>4.283750</td>\n",
       "      <td>1210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSWIM6_1</th>\n",
       "      <td>TGTATATA</td>\n",
       "      <td>916</td>\n",
       "      <td>1855</td>\n",
       "      <td>1.821500</td>\n",
       "      <td>939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZWILCH_0</th>\n",
       "      <td>TGTAAATA</td>\n",
       "      <td>470</td>\n",
       "      <td>1160</td>\n",
       "      <td>0.332625</td>\n",
       "      <td>690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZWILCH_1</th>\n",
       "      <td>TGTAAAAA</td>\n",
       "      <td>1021</td>\n",
       "      <td>1160</td>\n",
       "      <td>-0.272875</td>\n",
       "      <td>139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDA</th>\n",
       "      <td>TGTACATT</td>\n",
       "      <td>1636</td>\n",
       "      <td>2591</td>\n",
       "      <td>0.172625</td>\n",
       "      <td>955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDB</th>\n",
       "      <td>TGTACATT</td>\n",
       "      <td>2399</td>\n",
       "      <td>3017</td>\n",
       "      <td>-0.201500</td>\n",
       "      <td>618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B_0</th>\n",
       "      <td>TGTAAATT</td>\n",
       "      <td>396</td>\n",
       "      <td>5712</td>\n",
       "      <td>0.546250</td>\n",
       "      <td>5316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B_1</th>\n",
       "      <td>TGTATAAA</td>\n",
       "      <td>416</td>\n",
       "      <td>5712</td>\n",
       "      <td>1.155500</td>\n",
       "      <td>5296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B_2</th>\n",
       "      <td>TGTATATT</td>\n",
       "      <td>444</td>\n",
       "      <td>5712</td>\n",
       "      <td>0.440250</td>\n",
       "      <td>5268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B_3</th>\n",
       "      <td>TGTATATT</td>\n",
       "      <td>5521</td>\n",
       "      <td>5712</td>\n",
       "      <td>1.268125</td>\n",
       "      <td>191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1_0</th>\n",
       "      <td>TGTAAATA</td>\n",
       "      <td>1932</td>\n",
       "      <td>2444</td>\n",
       "      <td>0.304625</td>\n",
       "      <td>512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1_1</th>\n",
       "      <td>TGTACAAA</td>\n",
       "      <td>2224</td>\n",
       "      <td>2444</td>\n",
       "      <td>1.308250</td>\n",
       "      <td>220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1_2</th>\n",
       "      <td>TGTAAAAA</td>\n",
       "      <td>2368</td>\n",
       "      <td>2444</td>\n",
       "      <td>1.944125</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1_3</th>\n",
       "      <td>TGTACAAA</td>\n",
       "      <td>2421</td>\n",
       "      <td>2444</td>\n",
       "      <td>2.818875</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>TGTACAAA</td>\n",
       "      <td>714</td>\n",
       "      <td>1134</td>\n",
       "      <td>3.505125</td>\n",
       "      <td>420</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13337 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          PUM motif  position  UTR length    phyloP  distance\n",
       "42434.0    TGTAAATT       469        2757  0.816625      2288\n",
       "42615.0    TGTAAAAA      1785        2089 -0.185750       304\n",
       "42620.0    TGTAAAAA       977        1238  2.995500       261\n",
       "42622.0    TGTATAAA      1872        1938  2.347500        66\n",
       "42623.0    TGTATAAA       354        1470 -0.187000      1116\n",
       "42628.0    TGTAAAAA       848         901 -0.297750        53\n",
       "42430.0_0  TGTATATA      1928        4450  1.293000      2522\n",
       "42430.0_1  TGTATAAA      2056        4450  0.122375      2394\n",
       "42430.0_2  TGTATATT      2115        4450  0.320750      2335\n",
       "42430.0_3  TGTATATT      4002        4450  2.536500       448\n",
       "42435.0_0  TGTAAATA       209        4354  2.247125      4145\n",
       "42435.0_1  TGTATATT       282        4354  3.640250      4072\n",
       "42435.0_2  TGTACATT       440        4354  0.154875      3914\n",
       "42435.0_3  TGTACATT       596        4354  2.631625      3758\n",
       "42435.0_4  TGTATATA      4071        4354  0.107625       283\n",
       "42436.0_0  TGTATATA       142        1241  0.847875      1099\n",
       "42436.0_1  TGTATAAA       616        1241  0.847500       625\n",
       "42436.0_2  TGTAAATT       714        1241  1.211750       527\n",
       "42621.0_0  TGTATAAA       619        2815  1.121125      2196\n",
       "42621.0_1  TGTATATA       976        2815 -0.286875      1839\n",
       "42627.0_0  TGTATATT       113        2368  0.196375      2255\n",
       "42627.0_1  TGTATATT      2106        2368  0.037750       262\n",
       "A1CF_0     TGTACATT      1760        7320  0.046000      5560\n",
       "A1CF_1     TGTAAATA      4226        7320  0.773000      3094\n",
       "A1CF_2     TGTATATA      4334        7320  0.123500      2986\n",
       "A1CF_3     TGTAAATT      6695        7320  0.447250       625\n",
       "AADACL3    TGTACATA      2444        2763  0.003750       319\n",
       "AADAT      TGTAAAAA       407         705  0.190875       298\n",
       "AAED1      TGTATATA       438         507  0.459625        69\n",
       "AAK1_0     TGTAAATT       972       17874  2.640000     16902\n",
       "...             ...       ...         ...       ...       ...\n",
       "ZRANB2     TGTAAATA      1817        1888  2.950750        71\n",
       "ZRANB3     TGTAAATA      1377        3373  0.303375      1996\n",
       "ZSCAN12    TGTAAATT       320        1503  0.400500      1183\n",
       "ZSCAN20    TGTATATA        35         919 -0.112500       884\n",
       "ZSCAN23    TGTAAATA      1572        1810  0.383750       238\n",
       "ZSCAN25    TGTAAATA      1620        2387  0.473125       767\n",
       "ZSCAN30    TGTATATA      1174        2392 -0.312250      1218\n",
       "ZSCAN4_0   TGTATAAA        19         247 -0.209250       228\n",
       "ZSCAN4_1   TGTATATT        41         247 -0.239000       206\n",
       "ZSWIM1     TGTATATT       273        1216  0.265000       943\n",
       "ZSWIM2     TGTATATA        88         504  0.219375       416\n",
       "ZSWIM4     TGTATATT       998        1180  1.156625       182\n",
       "ZSWIM5_0   TGTAAATA       371        2050  2.706125      1679\n",
       "ZSWIM5_1   TGTACATA       900        2050 -0.100125      1150\n",
       "ZSWIM5_2   TGTATAAA      1121        2050  0.233500       929\n",
       "ZSWIM6_0   TGTATATA       645        1855  4.283750      1210\n",
       "ZSWIM6_1   TGTATATA       916        1855  1.821500       939\n",
       "ZWILCH_0   TGTAAATA       470        1160  0.332625       690\n",
       "ZWILCH_1   TGTAAAAA      1021        1160 -0.272875       139\n",
       "ZXDA       TGTACATT      1636        2591  0.172625       955\n",
       "ZXDB       TGTACATT      2399        3017 -0.201500       618\n",
       "ZYG11B_0   TGTAAATT       396        5712  0.546250      5316\n",
       "ZYG11B_1   TGTATAAA       416        5712  1.155500      5296\n",
       "ZYG11B_2   TGTATATT       444        5712  0.440250      5268\n",
       "ZYG11B_3   TGTATATT      5521        5712  1.268125       191\n",
       "ZZEF1_0    TGTAAATA      1932        2444  0.304625       512\n",
       "ZZEF1_1    TGTACAAA      2224        2444  1.308250       220\n",
       "ZZEF1_2    TGTAAAAA      2368        2444  1.944125        76\n",
       "ZZEF1_3    TGTACAAA      2421        2444  2.818875        23\n",
       "ZZZ3       TGTACAAA       714        1134  3.505125       420\n",
       "\n",
       "[13337 rows x 5 columns]"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sep = pandas.DataFrame.from_dict(dic, orient = 'index')\n",
    "df_sep.columns = ['PUM motif', 'position', 'UTR length', 'phyloP']\n",
    "\n",
    "df_sep['distance'] = df_sep['UTR length'] - df_sep['position']\n",
    "\n",
    "df_sep.sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df_sep_high = df_sep[df_sep['phyloP'] >1]\n",
    "df_sep_low = df_sep[df_sep['phyloP'] <1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "plt.hist(df_sep_high['distance'], bins = 3000, color = 'r', alpha = 0.5, label = 'Conserved PUM motif')\n",
    "plt.hist(df_sep_low['distance'], bins = 1000, color = 'gray', alpha = 0.5, label = 'unconserved PUM motif')\n",
    "plt.xlabel('Distance from 3end (bp)')\n",
    "plt.ylabel('Number of PUM motif')\n",
    "plt.legend()\n",
    "plt.xlim(0, 1000)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "dfTGTAAAAA = df_sep[df_sep['PUM motif'] == 'TGTAAAAA']\n",
    "dfTGTAAAAT = df_sep[df_sep['PUM motif'] == 'TGTAAAAT']\n",
    "dfTGTAAATA = df_sep[df_sep['PUM motif'] == 'TGTAAATA']\n",
    "dfTGTAAATT = df_sep[df_sep['PUM motif'] == 'TGTAAATT']\n",
    "\n",
    "dfTGTATAAA = df_sep[df_sep['PUM motif'] == 'TGTATAAA']\n",
    "dfTGTATAAT = df_sep[df_sep['PUM motif'] == 'TGTATAAT']\n",
    "dfTGTATATA = df_sep[df_sep['PUM motif'] == 'TGTATATA']\n",
    "dfTGTATATT = df_sep[df_sep['PUM motif'] == 'TGTATATT']\n",
    "\n",
    "dfTGTACAAA = df_sep[df_sep['PUM motif'] == 'TGTACAAA']\n",
    "dfTGTACAAT = df_sep[df_sep['PUM motif'] == 'TGTACAAT']\n",
    "dfTGTACATA = df_sep[df_sep['PUM motif'] == 'TGTACATA']\n",
    "dfTGTACATT = df_sep[df_sep['PUM motif'] == 'TGTACATT']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df_TGTAAAAA = dfTGTAAAAA['phyloP'] \n",
    "df_TGTAAAAT = dfTGTAAAAT['phyloP']\n",
    "df_TGTAAATA = dfTGTAAATA['phyloP']\n",
    "df_TGTAAATT = dfTGTAAATT['phyloP']\n",
    "\n",
    "df_TGTATAAA = dfTGTATAAA['phyloP']\n",
    "df_TGTATAAT = dfTGTATAAT['phyloP']\n",
    "df_TGTATATA = dfTGTATATA['phyloP']\n",
    "df_TGTATATT = dfTGTATATT['phyloP']\n",
    "\n",
    "df_TGTACAAA = dfTGTACAAA['phyloP']\n",
    "df_TGTACAAT = dfTGTACAAT['phyloP']\n",
    "df_TGTACATA = dfTGTACATA['phyloP']\n",
    "df_TGTACATT = dfTGTACATT['phyloP']\n",
    "\n",
    "TGTAAAAA = df_TGTAAAAA.as_matrix()\n",
    "TGTAAAAT = df_TGTAAAAT.as_matrix()\n",
    "TGTAAATA = df_TGTAAATA.as_matrix()\n",
    "TGTAAATT = df_TGTAAATT.as_matrix()\n",
    "\n",
    "TGTATAAA = df_TGTATAAA.as_matrix()\n",
    "TGTATAAT = df_TGTATAAT.as_matrix()\n",
    "TGTATATA = df_TGTATATA.as_matrix()\n",
    "TGTATATT = df_TGTATATT.as_matrix()\n",
    "\n",
    "TGTACAAA = df_TGTACAAA.as_matrix()\n",
    "TGTACAAT = df_TGTACAAT.as_matrix()\n",
    "TGTACATA = df_TGTACATA.as_matrix()\n",
    "TGTACATT = df_TGTACATT.as_matrix()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "data_to_plot = [TGTATATA, TGTAAATA, TGTACATA, TGTATAAA, TGTAAAAA, TGTACAAA, TGTATATT, TGTAAATT, TGTACATT]\n",
    "\n",
    "fig = plt.figure()\n",
    "ax = fig.add_subplot(111)\n",
    "\n",
    "ax.boxplot(data_to_plot, notch = True)\n",
    "\n",
    "# customize the whiskers, caps, fliers and medians\n",
    "plt.ylabel('Conservation_score (phyloP)')   \n",
    "plt.xticks([1,2,3,4,5,6,7,8,9], ['TGTATATA', 'TGTAAATA', 'TGTACATA', 'TGTATAAA', 'TGTAAAAA', 'TGTACAAA', 'TGTATATT', 'TGTAAATT', 'TGTACATT'])\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1921\n",
      "2249\n",
      "1212\n",
      "1327\n",
      "1632\n",
      "857\n",
      "1699\n",
      "1356\n",
      "1051\n"
     ]
    }
   ],
   "source": [
    "for i in data_to_plot:\n",
    "    print np.count_nonzero(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
